Fix delete.term-replacement to only check requested type/term exists

scott-christopher opened the pull request:
## Overview

Resolves #1804

Only checks that the requested type/term exists, then utilise misses and unrequested type/term for error messaging only.

## Implementation notes

This changes to only make use of the hits for the requested term/type, rather than previously checking for the absence of misses.

A new output message is also introduced that makes use of the misses and unrequested hits to report errors.

## Interesting/controversial decisions

n/a

## Test coverage

The existing test coverage is utilised.

From a comment by @dolio on Slack:
> * Try a delete for something that is neither a term nor a type. Should get a 'missing' error.
> * Try a delete for a term/type that is actually a type/term. That should give an error about finding the opposite thing.
> * Try a delete where there is a term and a type with the same name. That should no longer display an error.

It would be great to get some advice on how to get achieve and include this coverage, as it is currently lacking from this PR.

## Loose ends

See above note on test coverage.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unisonweb/unison/1810)
<!-- Reviewable:end -->

pchiusano commented:
Looks great! @scott-christopher 

For testing, I'd just add some stanzas to [this transcript](https://raw.githubusercontent.com/unisonweb/unison/trunk/unison-src/transcripts/deleteReplacements.md), see [this guide to the syntax of transcripts](https://raw.githubusercontent.com/unisonweb/unison/trunk/unison-src/transcripts/hello.md). You can use ` ```ucm:error ` blocks when errors are expected and you want the transcript output to include these errors.

Once that's in we can get this merged.

pchiusano reviewed (approved):
Great, thanks!

pchiusano commented:
Oops, I misread, it looks like those tests I suggested still haven't been added. If you are out of steam on this, let us know and we can merge and create a ticket to track that loose end. Otherwise, the additional testing coverage would be great! 🙏

scott-christopher commented:
Hi @pchiusano - sorry, I am still keen to finish this off, it's just been a struggle to find the time to set aside recently. I'll likely jump on Slack for some pointers too, as I was unable to produce examples that triggered some of the tests.

scott-christopher commented:
That said, if anyone else would like to jump on this in the interim then please don't wait for my sake.

scott-christopher committed:
Fix delete.term-replacement to only check requested type/term exists

Addresses #1804

Only checks that the requested type/term exists, then utilise misses and unrequested type/term for error messaging only

ChrisPenner commented on unison-cli/src/Unison/Codebase/Editor/Output.hs:
```diff
@@ -153,6 +153,7 @@ data Output
   | NoLastRunResult
   | SaveTermNameConflict Name
   | SearchTermsNotFound [HQ.HashQualified Name]
+  | SearchTermsNotFound' Bool [HQ.HashQualified Name] [HQ.HashQualified Name]
```
Some docs on this and how it's different from `SearchTermsNotFound` would be really great.

I would personally prefer a more descriptive name than `'` so we can see at a glance what the differentiator is 😄 

Also a note on what the `Bool` is, and what the difference between the two HQ args is would be nice 😄

ChrisPenner reviewed (commented):
This transcript didn't behave quite like I'd expect:

    ```unison
    bar = 3
    unique[aa] type bar = Foo
    ```

    ```ucm
    .> add
    ```

    ```unison
    unique[bb] type bar = Foo | Bar
    ```

    ```ucm
    .> update
    .> view.patch
    .> delete.type-replacement bar
    .> view.patch
    ```

Even though there's both a type and term for the name `bar` I'd expect that it would find the one I asked for and delete it.

Note that this works if you do `delete.type-replacement 1` which is hash-qualified.

Here's the output:

    ```unison
    unique[aa] type bar = Foo
    ```

    ```ucm

      I found and typechecked these definitions in scratch.u. If you
      do an `add` or `update`, here's how your codebase would
      change:
      
        ⍟ These new definitions are ok to `add`:
        
          unique type bar

    ```
    ```ucm
    .> add

      ⍟ I've added these definitions:
      
        unique type bar

    ```
    ```unison
    unique[bb] type bar = Foo | Bar
    ```

    ```ucm

      I found and typechecked these definitions in scratch.u. If you
      do an `add` or `update`, here's how your codebase would
      change:
      
        ⍟ These names already exist. You can `update` them to your
          new definition:
        
          unique type bar

    ```
    ```ucm
    .> update

      ⍟ I've updated these names to your new definition:
      
        unique type bar

    .> view.patch

      Edited Types: 1. bar#evhqg163jj -> 2. bar
      
      Tip: To remove entries from a patch, use
          delete.term-replacement or delete.type-replacement, as
          appropriate.

    .> delete.type-replacement bar

      Done.

    .> view.patch

      Edited Types: 1. bar#evhqg163jj -> 2. bar
      
      Tip: To remove entries from a patch, use
          delete.term-replacement or delete.type-replacement, as
          appropriate.

    ```

ChrisPenner commented:
Hrmm, actully this one very simple transcript doesn't work on this Pr or on trunk either:

    ```unison
    bar = 3
    ```

    ```ucm
    .> add
    ```

    ```unison
    bar  = 4
    ```

    ```ucm
    .> update
    .> view.patch
    .> delete.term-replacement bar
    .> view.patch
    ```

Maybe I'm misunderstanding how replacement deletion is meant to work, but it'd be nice to be able to specify a name right? Or does it only accept full hash-qualifications?

If the latter we should probably change the help-text as well as the recommendation within `view.patch` to indicate as such 🤔 

Also it probably shouldn't succeed with "Done." when it doesn't do anything I think.

Here's the transcript output for that one:

    ```unison
    bar = 3
    ```

    ```ucm

      I found and typechecked these definitions in scratch.u. If you
      do an `add` or `update`, here's how your codebase would
      change:
      
        ⍟ These new definitions are ok to `add`:
        
          bar : ##Nat

    ```
    ```ucm
    .> add

      ⍟ I've added these definitions:
      
        bar : ##Nat

    ```
    ```unison
    bar  = 4
    ```

    ```ucm

      I found and typechecked these definitions in scratch.u. If you
      do an `add` or `update`, here's how your codebase would
      change:
      
        ⍟ These names already exist. You can `update` them to your
          new definition:
        
          bar : ##Nat

    ```
    ```ucm
    .> update

      ⍟ I've updated these names to your new definition:
      
        bar : ##Nat

    .> view.patch

      Edited Terms: 1. bar#f3lgjvjqoo -> 2. bar
      
      Tip: To remove entries from a patch, use
          delete.term-replacement or delete.type-replacement, as
          appropriate.

    .> delete.term-replacement bar

      Done.

    .> view.patch

      Edited Terms: 1. bar#f3lgjvjqoo -> 2. bar
      
      Tip: To remove entries from a patch, use
          delete.term-replacement or delete.type-replacement, as
          appropriate.

    ```

ChrisPenner reviewed (approved):
Improved errors look good!
We should create another issue to track the behaviour I mentioned though I think 😄

aryairani commented:
Yeah @ChrisPenner Could you add a ticket to track.  We could also separately also add a transcript to trunk to demonstrate existing behavior, even if it's undesirable; the markdown can include a discussion.

aryairani reviewed (approved):
Thanks @scott-christopher for the PR and thanks @tstat for refreshening it. I agree it would be good to rename/document `SearchTermsNotFound'` if it's fresh in your mind; or else open another ticket to track and merge as-is.

ChrisPenner commented:
Created https://github.com/unisonweb/unison/issues/3366 to track

tstat committed:
rename SearchTermsNotFound' and document
